### PR TITLE
Reserved Events

### PR DESCRIPTION
**Pull request by sojanpr**
PR with following implementations

1) v2 Events for tracking custom  and standard events. Builders for  v2 event tracking

2) BUO changes for adding more metadata

4) Mote test routines for buo serialization -  deserialization

@EvangelosG @derrickstaten @dmitrig01

### CODE REVIEW

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java [hunk 1]
```diff
@@ -240,7 +262,7 @@ public void onClick(View view) {
                 LinkProperties linkProperties = new LinkProperties()
                         .addTag("myShareTag1")
                         .addTag("myShareTag2")
-                        //.setAlias("mylinkName") // In case you need to white label your link
+                                //.setAlias("mylinkName") // In case you need to white label your link
```
Comment (derrickstaten) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java]: any reason for the extra tab?

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java [hunk 2]
```diff
@@ -309,6 +331,40 @@ public String getSharingMessageForChannel(String channel) {
...
+                        .setTax(12.3)
+                        .setRevenue(1.5)
+                        .setCondition(BranchEvent.CONDITION.FAIR)
+                        .setTransactionID("12344555")
+                        .addCustomProperty("Custom_Event_Property_Key1", "Custom_Event_Property_val1")
+                        .addCustomProperty("Custom_Event_Property_Key2", "Custom_Event_Property_val2")
+                        .addContentItems(branchUniversalObject)
+                        .logEvent(MainActivity.this);
+
+
```
Comment (derrickstaten) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java]: extra whitespace?

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java [hunk 3]
```diff
@@ -309,6 +331,40 @@ public String getSharingMessageForChannel(String channel) {
...
+            public void onClick(View v) {
+                new BranchEvent(BRANCH_STANDARD_EVENT.PURCHASE)
+                        .setAffiliation("test_affiliation")
+                        .setCoupon("test_coupon")
+                        .setCurrency(CurrencyType.USD)
+                        .setDescription("Event _description")
+                        .setShipping(10.2)
+                        .setTax(12.3)
+                        .setRevenue(1.5)
+                        .setCondition(BranchEvent.CONDITION.FAIR)
```
Comment (derrickstaten) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java]: Whoa. @E-B-Smith are you doing something similar on iOS for this and other constants?

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java [hunk 4]
```diff
@@ -309,6 +331,40 @@ public String getSharingMessageForChannel(String channel) {
...
+                        .setTransactionID("12344555")
+                        .addCustomProperty("Custom_Event_Property_Key1", "Custom_Event_Property_val1")
+                        .addCustomProperty("Custom_Event_Property_Key2", "Custom_Event_Property_val2")
+                        .addContentItems(branchUniversalObject)
```
Comment (derrickstaten) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java]: so `.addContentItems` takes one item as opposed to a list/array?

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/test/BUOTestRoutines.java [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+        Parcel parcel = Parcel.obtain();
+        buo.writeToParcel(parcel, 0);
+        parcel.setDataPosition(0);
+        BranchUniversalObject buoCreatedFromParcel = (BranchUniversalObject) BranchUniversalObject.CREATOR.createFromParcel(parcel);
+
+        return (checkIfIdenticalJson(buo.getContentMetadata().convertToJson(), buo2.getContentMetadata().convertToJson()))
+                && (checkIfIdenticalJson(testBuoJson1, testBuoJson2))
+                && (checkIfIdenticalJson(testBuoJson1, buoCreatedFromParcel.convertToJson()));
+
+
```
Comment (derrickstaten) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/test/BUOTestRoutines.java]: lots of whitespace

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/test/BUOTestRoutines.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+                Log.e("BranchTestBed", "Error : BUO serialisation error.");
+            }
+        } else {
+            Log.e("BranchTestBed", "Serialised versions buo versions are not matching \n" + obj1 + "\n\n" + obj2);
+        }
+
+        return isIdentical;
+    }
+
+}
```
Comment (derrickstaten) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/test/BUOTestRoutines.java]: love the testing! I didn't look to see if it covers every single field. Let me know if you want help with that.

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 7]
```diff
@@ -315,6 +334,19 @@ public void userCompletedAction(String action, HashMap<String, String> metadata)
...
     /**
      * <p>
+     * Method to report user actions happened on this BUO. Use this method to report the user actions for analytics purpose.
+     * </p>
+     *
+     * @param action   A {@link BRANCH_STANDARD_EVENT }with value of user action name.  See {@link BRANCH_STANDARD_EVENT} for Branch defined user events.
+     * @param metadata A HashMap containing any additional metadata need to add to this user event
+     */
+    @SuppressWarnings("deprecation")
+    public void userCompletedAction(BRANCH_STANDARD_EVENT action, HashMap<String, String> metadata) {
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: Why are we adding a method, which is deprecated?

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 8]
```diff
@@ -429,9 +468,10 @@ public double getPrice() {
      * </p>
      *
      * @return {@link String} with ISO 4217 for this currency. Empty string if there is no currency type set
+     * @deprecated Please check {@link BranchEvent} for more info on commerce event tracking with Branch
      */
     public String getCurrencyType() {
-        return currency_.toString();
+        return null;
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: if we can return null here, why can't we return null for the above methods `getType()` and `getPrice()`? Sorry if this is a dumb/obvious answer.

File: Branch-SDK/src/io/branch/referral/Defines.java [hunk 9]
```diff
@@ -122,7 +123,46 @@
...
+        AddressCity("$address_city"),
+        AddressRegion("$address_region"),
+        AddressCountry("$address_country"),
+        AddressPostalCode("$address_postal_code"),
+        Latitude("$latitude"),
+        Longitude("$longitude"),
+        ImageCaptions("$image_captions"),
+        Condition("$condition"),
+        CreationTimestamp("$creation_timestamp"),
+        Runtime("$run_time");
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/Defines.java]: I think we can remove run_time, cc @zeeshabnc

File: Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+    SPEND_CREDITS("SPEND_CREDITS"),
+
+    // Content events
+    SEARCH("SEARCH"),
+    VIEW_CONTENT("VIEW_CONTENT"),
+    VIEW_CONTENT_LIST("VIEW_CONTENT_LIST"),
+    RATE("RATE"),
+    SHARE_CONTENT("SHARE_CONTENT"),
+
+    //Life cycle events
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java]: nitpick: add a space here

Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java]: also: `user lifecycle events`

File: Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
+package io.branch.referral.util;
+
+/**
+ * Created by sojanpr on 7/5/17.
+ * <p>
+ * Standard Branch Events enumeration. These events are used with Branch standard events
+ * </p>
+ */
+public enum BRANCH_STANDARD_EVENT {
+    // Commerce Events
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java]: nitpick: lowercase `Events` -> `events` to match the other two

File: Branch-SDK/src/io/branch/referral/util/BranchContentSchema.java [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+    MEDIA_VIDEO,
+    OTHER,
+    TEXT_ARTICLE,
+    TEXT_BLOG,
+    TEXT_OTHER,
+    TEXT_RECIPE,
+    TEXT_REVIEW,
+    TEXT_SEARCH_RESULTS,
+    TEXT_STORY,
+    TEXT_TECHNICAL_DOC;
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/util/BranchContentSchema.java]: @funeric555 this is for you :)

**Comment by derrickstaten** [comment]
I just left comments. Let me know if you want a more thorough review of any individual pieces.

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java (hunk 4 above)
Comment (sojanpr) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/MainActivity.java]: There are  2 methods  
`addContentItems(BranchUniversalObject... contentItems)`
`addContentItems(List<BranchUniversalObject> contentItems)` 
Both can take multiple params

File: Branch-SDK-TestBed/src/io/branch/branchandroiddemo/test/BUOTestRoutines.java (hunk 6 above)
Comment (sojanpr) [inline Branch-SDK-TestBed/src/io/branch/branchandroiddemo/test/BUOTestRoutines.java]: This test serialize and deserialize and then compare . Tests both Json serialisataion and Parcelable serialization. Each individual  fields are checked  by iterating through the BUO elements. So it covers all the fields

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java (hunk 7 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: We are  still maintaining deprecated method
`public void userCompletedAction(String action, HashMap<String, String> metadata)`  
for backward compatibility. After a while we will make this method private and then remove deprecation

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java (hunk 8 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: Good catch.` getType` also can return null for consistency. Will change

File: Branch-SDK/src/io/branch/referral/Defines.java (hunk 9 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/Defines.java]: Yes we don't use it. Removed  from const

File: Branch-SDK/src/io/branch/referral/util/BranchContentSchema.java (hunk 12 above)
Comment (sojanpr) [inline Branch-SDK/src/io/branch/referral/util/BranchContentSchema.java]: Will fix it :)

File: Branch-SDK/src/io/branch/referral/DeviceInfo.java [hunk 13]
```diff
@@ -159,6 +159,49 @@ public void updateRequestWithDeviceParams(JSONObject requestObj) {
...
     /**
+     * Update the given server request JSON with user data. Used for V2 events
+     *
+     * @param requestObj JSON object for Branch server request
+     */
+    public void updateRequestWithUserData(JSONObject requestObj) {
+        try {
+            if (!hardwareID_.equals(SystemObserver.BLANK)) {
+                requestObj.put(Defines.Jsonkey.AndroidID.getKey(), hardwareID_);
+                requestObj.put(Defines.Jsonkey.IsHardwareIDReal.getKey(), isHardwareIDReal_);
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/DeviceInfo.java]: We don't want any fake Android IDs, and no need to send is_hardware_id_real.

File: Branch-SDK/src/io/branch/referral/DeviceInfo.java [hunk 14]
```diff
@@ -159,6 +159,49 @@ public void updateRequestWithDeviceParams(JSONObject requestObj) {
...
+            if (!brandName_.equals(SystemObserver.BLANK)) {
+                requestObj.put(Defines.Jsonkey.Brand.getKey(), brandName_);
+            }
+            if (!modelName_.equals(SystemObserver.BLANK)) {
+                requestObj.put(Defines.Jsonkey.Model.getKey(), modelName_);
+            }
+            requestObj.put(Defines.Jsonkey.ScreenDpi.getKey(), screenDensity_);
+            requestObj.put(Defines.Jsonkey.ScreenHeight.getKey(), screenHeight_);
+            requestObj.put(Defines.Jsonkey.ScreenWidth.getKey(), screenWidth_);
+            requestObj.put(Defines.Jsonkey.WiFi.getKey(), isWifiConnected_);
```
Comment (derrickstaten) [inline Branch-SDK/src/io/branch/referral/DeviceInfo.java]: we don't need wifi

**Comment by derrickstaten** [comment]
user_data fields that seem to be missing (though I do not understand the SDK code):
- environment
- user_agent (if available)
- device_fingerprint_id
- sdk
- app_version

File: Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    VIEW_ITEM("VIEW_ITEM"),
+    VIEW_ITEMS("VIEW_ITEMS"),
+    RATE("RATE"),
+    SHARE("SHARE"),
+
+    // User lifecycle events
+    COMPLETE_REGISTRATION("COMPLETE_REGISTRATION"),
+    COMPLETE_TUTORIAL("COMPLETE_TUTORIAL"),
+    ACHIEVE_LEVEL("ACHIEVE_LEVEL"),
+    UNLOCK_ACHIEVEMENTS("UNLOCK_ACHIEVEMENT");
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java]: UNLOCK_ACHIEVEMENT, not UNLOCK_ACHIEVEMENTS

File: Branch-SDK/src/io/branch/referral/util/BranchEvent.java [hunk 16]
```diff
@@ -1,21 +1,303 @@
...
+
+    /**
+     * Set the condition of the content item. Value is one of the enum constants from {@link BranchEvent.CONDITION}
+     *
+     * @param condition {@link BranchEvent.CONDITION}
+     * @return This object for chaining builder methods
+     */
+    public BranchEvent setCondition(CONDITION condition) {
+        return addStandardProperty(Defines.Jsonkey.Condition.getKey(), condition.name());
+    }
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/BranchEvent.java]: this belongs on contentmetadata, not on the branchevent

File: Branch-SDK/src/io/branch/referral/util/BranchEvent.java [hunk 17]
```diff
@@ -1,21 +1,303 @@
...
+    /**
+     * Set description for this transaction event
+     *
+     * @param description {@link String } transaction description
+     * @return This object for chaining builder methods
+     */
+    public BranchEvent setDescription(String description) {
+        return addStandardProperty(Defines.Jsonkey.Description.getKey(), description);
+    }
+
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/BranchEvent.java]: missing searchQuery

File: Branch-SDK/src/io/branch/referral/util/ContentMetadata.java [hunk 18]
```diff
@@ -0,0 +1,415 @@
...
+     */
+    public Double quantity;
+    /**
+     * Any price associated with the qualifying content item
+     */
+    public Double price;
+    /**
+     * Currency type associated with the price
+     */
+    public CurrencyType currencyType;
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/ContentMetadata.java]: this belongs in the event data

File: Branch-SDK/src/io/branch/referral/util/ContentMetadata.java [hunk 19]
```diff
@@ -0,0 +1,415 @@
...
+     */
+    public Double averageRating;
+    /**
+     * Total number of ratings for the qualifying content item
+     */
+    public Integer ratingCount;
+    /**
+     * Maximum ratings for the qualifying content item
+     */
+    public Double maximumRating;
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/ContentMetadata.java]: can we call it ratingMax

File: Branch-SDK/src/io/branch/referral/util/ContentMetadata.java [hunk 20]
```diff
@@ -0,0 +1,415 @@
...
+     */
+    public ProductCategory productCategory;
+    /**
+     * Variant of product if this metadata is for a product
+     */
+    public String productVariant;
+    /**
+     * Average rating for the qualifying content item
+     */
+    public Double averageRating;
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/ContentMetadata.java]: can we call it ratingAverage

File: Branch-SDK/src/io/branch/referral/util/ContentMetadata.java [hunk 21]
```diff
@@ -0,0 +1,415 @@
...
+     */
+    public Double maximumRating;
+    /**
+     * Street address associated with the qualifying content item
+     */
+    public String addressStreet;
+    /**
+     * City name associated with the qualifying content item
+     */
+    private String addressCity;
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/ContentMetadata.java]: why is this one private when all others are public?

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 22]
```diff
@@ -324,14 +353,24 @@ public boolean isPublicallyIndexable() {
         return indexMode_ == CONTENT_INDEX_MODE.PUBLIC;
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: fyi, I know this is already an existing function, bht publically is not correct spelling - the correct spelling (and it's this way in v2/event) is publicly

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java [hunk 23]
```diff
@@ -324,14 +353,24 @@ public boolean isPublicallyIndexable() {
...
     }
 
+    private boolean isLocallyIndexable() {
+        return localIndexMode_ == CONTENT_INDEX_MODE.PUBLIC;
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: seems wrong logic?

Comment (sojanpr) [reply Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: isLocallyIndexable returns true only if content indexing mode is private.

Comment (dmitrig01) [reply Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: whoops yes, sorry, I didn't read the _local_ indexmode

File: Branch-SDK/src/io/branch/referral/util/BranchEvent.java [hunk 24]
```diff
@@ -1,21 +1,303 @@
...
  * </p>
  */
 public class BranchEvent {
+    private final String eventName;
+    private final boolean isStandardEvent;
+    private final JSONObject standardProperties;
+    private final JSONObject customProperties;
+    private final List<BranchUniversalObject> buoList;
+
+    public enum CONDITION {OTHER, NEW, GOOD, FAIR, POOR, USED, REFURBISHED}
```
Comment (dmitrig01) [inline Branch-SDK/src/io/branch/referral/util/BranchEvent.java]: missed EXCELLENT

File: Branch-SDK/src/io/branch/referral/util/BranchEvent.java (hunk 16 above)
Comment (sojanpr) [reply Branch-SDK/src/io/branch/referral/util/BranchEvent.java]: Agree that it make sense to have it in ContentMetadata. Is our back end is ready to handle it if we move this to ContentMetadata

File: Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java (hunk 15 above)
Comment (sojanpr) [reply Branch-SDK/src/io/branch/referral/util/BRANCH_STANDARD_EVENT.java]: UNLOCK_ACHIEVEMENT ? or you mean to say UNLOCK_ACHIEVEMENTS

File: Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java (hunk 22 above)
Comment (sojanpr) [reply Branch-SDK/src/io/branch/indexing/BranchUniversalObject.java]: Will make it with next API onwards

File: Branch-SDK/src/io/branch/referral/util/ContentMetadata.java (hunk 18 above)
Comment (sojanpr) [reply Branch-SDK/src/io/branch/referral/util/ContentMetadata.java]: We have it with Branch Event

**Comment by dmitrig01** [comment]
Hey Sojan, sorry about the delay in review here. I have a final review:

- let's add AAID (google advertising id) to user_info. and fix unidentified_device to mean it has neither android_id nor aaid
- add environment (instant_app vs full_app)
- add limit_ad_tracking: true/false
- rename ratingAvg to ratingAverage
- productCondition should be called condition
- currencyType should be called currency
- for $custom_fields, we should just make that top level within the BUO - not have a custom field for it
- for the product categories renames - (food & bev / cameras & optics) -- can you change the constants as well?

cc @zeeshabnc @derrickstaten

**Comment by dmitrig01** [comment]
Oh, one more thing:
- can we split the "sdk" field into the "android" part (in "sdk") and the version, in "sdk_version"

**Comment by sojanpr** [comment]
@dmitri Thanks for the review.  Few points for clarification
1) You mean to move`AAID` and `limit_ad_tracking` (`limit_ad_tracking` is already there along with the `AAID`)   to `user_info` right?
2) `environment` is already part of the `user_data`
3) `CurrencyType` is an already existing Enum in SDK. the key the post body is `currency`
4) FOOD_BEVERAGE_AND_TOBACCO -> FOOD&BEV ?

**Comment by dmitrig01** [comment]
1. yes, exactly. perhaps i missed them...
2. ah, ust have missed it as well
3. ok. the field in the buo is called `CurrencyType currencyType` can we make it `CurrencyType currency`?
4. no, FOOD_BEVERAGES_AND_TOBACCO - so it's consistent with the string value

File: README.md [hunk 25]
```diff
@@ -563,7 +563,36 @@ The universal object is where you define all of the custom metadata associated w
 
 **expirationDate**: The date when the content will not longer be available or valid. Currently, this is only used for Spotlight indexing but will be used by Branch in the future.
 
+
+### Tracking User Actions and Events
+Use `BranchEvent` class to track special user actions or application specific events beyond app installs, opens, and sharing . You can track events such as when a user adds an item to an on-line shopping cart, or searches for a keyword etc.
+`BranchEvent` provides interface to add contents represented by BranchUniversalObject in order to associate contents with events.
```
Comment (zeeshabnc) [inline README.md]: BranchEvent provides an interface to add content(s) represented by a BranchUniversalObject in order to associated content(s) with events. 

(Added *an* interface, *a* BranchUniversalObject)

File: README.md [hunk 26]
```diff
@@ -563,7 +563,36 @@ The universal object is where you define all of the custom metadata associated w
...
+Analytics about your app's BranchEvents can be found on the Branch dashboard, and BranchEvents also provide tight integration with many third party analytics providers.
```
Comment (zeeshabnc) [inline README.md]: Sentence is too long. Maybe split it up into two sentences. 

You can view analytics for the BranchEvents you fire on the Branch dashboard. BranchEvents provide a seamless integration with many third party analytics providers like <X,Y>
@Sarkar  what should we say X,Y are?

Comment (Sarkar) [reply README.md]: Google Analytics, Criteo

**Pull request merged**